types/node/zlib.d.ts

```diff
@@ -110,9 +110,15 @@ declare module "zlib" {
         memLevel?: number | undefined; // compression only
         strategy?: number | undefined; // compression only
         dictionary?: NodeJS.ArrayBufferView | ArrayBuffer | undefined; // deflate/inflate only, empty dictionary by default
-        info?: boolean | undefined;
         maxOutputLength?: number | undefined;
     }
+    interface ZlibOptionsWithoutInfo extends ZlibOptionsBase {
+        info?: false | undefined;
+    }
+    interface ZlibOptionsWithInfo extends ZlibOptionsBase {
+        info: true;
+    }
+    type ZlibOptions = ZlibOptionsWithoutInfo | ZlibOptionsWithInfo;
```

Comment:
I would beg to differ because leaving `ZlibOptions` as is would mean that the overloads I made won't work. Because when the `info` property of the `ZlibOptions` stays the same, `ZlibOptionsWithInfo` would remain backwards-compatible which won't play well with the overloads. The lack of the (correct) implementation of the `info` property unfortunately means that a fix would be impactful. At the end of the day you unfortunately have to work with overloads as the functions of the `Zlib` library have 2 different return types based on the `info` property of the `ZlibOptions`.

Could you see my point of view?

Reply:
TypeScript picks the first matching overload. If the overload using `ZlibOptionsWithInfo` goes first, it takes precedence and the callback / return values are inferred the way I think you intend to happen.

You can see this in the playground link if you hover the `result` arguments in the callbacks.

Reply:
I don't think that is a very solid design strategy. Not to mention that I'd doubt it would be much more less impactful because it would remain a breaking change because the return type does still change from `Buffer` to the info object when `info` equals `true`.

Reply:
I have made a change, `ZlibOptions` interface has been reinstated in full glory and it no longer is a union type of my two newly introduced interfaces. Changes to the `@types/inflation` package has been reverted by this as well and the package is now unaffected.

Reply:
Putting the more specific overload first is the normal way to write overloads in TypeScript; what's not a solid design strategy? I would think there's precedent when it comes to say, `child_process`'s return values that could be used to model the API here.

Reply:
Okay, thanks for the clarification on that part. The main take away would be that it doesn't prevent this from being a breaking change though.